hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/ServerCall.java

```diff
@@ -175,23 +178,17 @@ private void release(int mask) {
     }
   }
 
-  @Override
-  public void cleanup() {
-    release(0b01);
+  public void retainByWAL() {
+    reference.incrementAndGet();
   }
 
-  public void retainByWAL() {
-    for (;;) {
-      int ref = reference.get();
-      int nextRef = ref | 0b10;
-      if (reference.compareAndSet(ref, nextRef)) {
-        return;
+  public void releaseByWAL() {
```

Comment:
I saw the review request is closed. Let me post one comment here. I think there is one corner case here.
batch() -> FSWEntry (increase refCount)
consume() -> sync done, decrease refCount -> cleanup()
next FSWEntry(increase refCount) --> referring to the buffer which has been released.

In this case, it is better to pre-calculate how many refCount is there and set it at the very begin.

Reply:
As explained by Anoop, this is not a concern as FSWEntry() is called within handler thread, so bit for Call still holds, no cleanup will be called.